lib/loopback-iterator.js

```diff
@@ -0,0 +1,167 @@
+'use strict'
+
+const debug = require('debug')('loopback:mixin:iterator')
+const utils = require('loopback-datasource-juggler/lib/utils')
+const iterators = require('async-iterators')
+const async = require('async')
+const _ = require('lodash')
+
+module.exports = class LoopbackIterator {
+  constructor(Model, query, options = {}) {
+    this.Model = Model
+    this.query = query || {}
+
+    _.assign(this, _.defaults(options, {
+      batchSize: 100,
+      maxQueueLength: 50,
+      queueWaitInterval: 100,
+      concurrentItems: 50,
+    }))
+
+    this.itemsFrom = this.query.skip || 0
+    this.itemsTo = this.itemsFrom
+    this.limit = this.query.limit
+
+    this.initialized = false
+    this.itemsTotal = null
+    this.pageTotal = null
+    this.currentItem = 0
+    this.currentItems = []
+
+    this.queue = null
+  }
+
+  initialize(cb = utils.createPromiseCallback()) {
+    if (this.initialized) {
+      process.nextTick(() => cb())
+      return cb.promise
+    }
+
+    debug('Initializing')
+
+    const countWhere = this.query.where || {}
+
+    this.Model.count(countWhere)
+      .then(count => {
+        this.itemsTotal = this.limit ? Math.min(count, this.limit) : count
+        this.pageTotal = Math.ceil(this.itemsTotal / this.batchSize)
+        this.initialized = true
+        cb(null)
```

Comment:
Strictly speaking, I think the general best practice here is to callback with an explicit `null` instead of an implicit`undefined`. That is what is done in node core. I think functions that return `undefined` are generally frowned upon although for most actual use cases it doesn't really make much difference as generally the callers check for a falsy value, and not specifically `null` or `undefined`. I'll change it here to undefined here though since that's what we tend to use in all the rest of our code.